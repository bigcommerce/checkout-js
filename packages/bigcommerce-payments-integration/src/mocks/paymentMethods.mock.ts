export function getBigCommercePaymentsPaypalMethod() {
    return {
        clientToken: undefined,
        config: {
            cardCode: true,
            displayName: 'Credit Cards',
            enablePaypal: undefined,
            hasDefaultStoredInstrument: false,
            helpText: '',
            is3dsEnabled: undefined,
            isHostedFormEnabled: false,
            isVaultingCvvEnabled: undefined,
            isVaultingEnabled: false,
            isVisaCheckoutEnabled: undefined,
            logo: undefined,
            merchantId: '',
            requireCustomerCode: false,
            testMode: true,
        },
        gateway: undefined,
        id: 'bigcommerce_payments_paypal',
        initializationData: {
            attributionId: 'BC_id',
            availableAlternativePaymentMethods: [],
            buttonStyle: {
                color: 'black',
                shape: 'rect',
                height: 55,
                label: 'pay',
            },
            clientId: 'asd123',
            enabledAlternativePaymentMethods: [],
            intent: 'capture',
            isComplete: false,
            isDeveloperModeApplicable: false,
            isHostedCheckoutEnabled: true,
            isInlineCheckoutEnabled: false,
            isPayPalCreditAvailable: true,
            isVenmoEnabled: true,
            merchantId: 'XdasdQWe1123',
            orderId: undefined,
            shouldRenderFields: true,
            showOnlyOnMobileDevices: false,
        },
        initializationStrategy: {
            type: 'not_applicable',
        },
        logoUrl: 'http://logo_url_path',
        method: 'paypal',
        nonce: undefined,
        returnUrl: undefined,
        supportedCards: [
            'PAYPAL',
            'PAYPALCREDIT',
            'VISA',
            'MC',
            'AMEX',
            'DISCOVER',
            'VENMO',
            'APPLEPAY',
        ],
        type: 'PAYMENT_TYPE_API',
    };
}

export function getBigCommercePaymentsFastlaneMethod() {
    const bigCommercePayments = getBigCommercePaymentsPaypalMethod();

    return {
        ...bigCommercePayments,
        id: 'bigcommerce_payments_paylater',
        initializationData: {
            ...bigCommercePayments.initializationData,
            payPalCreditProductBrandName: {
                credit: 'Pay Later',
                messaging: 'Pay Later',
            },
        },
        method: 'paypal-credit',
    };
}

export function getBigCommercePaymentsVenmoMethod() {
    const bigCommercePayments = getBigCommercePaymentsPaypalMethod();

    return {
        ...bigCommercePayments,
        id: 'bigcommerce_payments_venmo',
        method: 'paypal-venmo',
    };
}

export function getBigCommercePaymentsAPMsMethod() {
    const bigcommercePayments = getBigCommercePaymentsPaypalMethod();

    const apms = ['bancontact', 'eps', 'giropay', 'ideal', 'mybank', 'p24', 'sepa', 'sofort'];

    return {
        ...bigcommercePayments,
        config: {
            ...bigcommercePayments.config,
            displayName: 'Sepa',
        },
        gateway: 'bigcommerce_payments_apms',
        id: 'sepa',
        initializationData: {
            ...bigcommercePayments.initializationData,
            availableAlternativePaymentMethods: apms,
            enabledAlternativePaymentMethods: apms,
        },
        method: 'multi-option',
    };
}
