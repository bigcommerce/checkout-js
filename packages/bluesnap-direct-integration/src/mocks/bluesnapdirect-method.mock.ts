import { PaymentMethod } from '@bigcommerce/checkout-sdk';

const DISPLAY_NAME_MAP = {
    credit_card: 'Credit Card',
    ecp: 'Electronic Check Presentment',
};

export function getBlueSnapDirect(
    method: keyof typeof DISPLAY_NAME_MAP = 'credit_card',
): PaymentMethod {
    return {
        clientToken: method === 'credit_card' ? 'pfToken' : '',
        config: {
            cardCode: true,
            displayName: DISPLAY_NAME_MAP[method],
            isHostedFormEnabled: true,
            testMode: true,
        },
        gateway: 'bluesnapdirect',
        id: method,
        method,
        supportedCards:
            method === 'credit_card'
                ? ['AMEX', 'CUP', 'DINERS', 'DISCOVER', 'JCB', 'MC', 'VISA']
                : [],
        type: 'PAYMENT_TYPE_API',
    };
}
