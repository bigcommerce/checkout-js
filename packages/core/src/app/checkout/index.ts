export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
