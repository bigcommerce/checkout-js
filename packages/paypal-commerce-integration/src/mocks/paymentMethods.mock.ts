export function getPayPalCommerceMethod() {
    return {
        clientToken: undefined,
        config: {
            cardCode: true,
            displayName: 'Credit Cards',
            enablePaypal: undefined,
            hasDefaultStoredInstrument: false,
            helpText: '',
            is3dsEnabled: undefined,
            isHostedFormEnabled: false,
            isVaultingCvvEnabled: undefined,
            isVaultingEnabled: false,
            isVisaCheckoutEnabled: undefined,
            logo: undefined,
            merchantId: '',
            requireCustomerCode: false,
            testMode: true,
        },
        gateway: undefined,
        id: 'paypalcommerce',
        initializationData: {
            attributionId: 'BC_id',
            availableAlternativePaymentMethods: [],
            buttonStyle: {
                color: 'black',
                shape: 'rect',
                height: 55,
                label: 'pay',
            },
            clientId: 'asd123',
            enabledAlternativePaymentMethods: [],
            intent: 'capture',
            isComplete: false,
            isDeveloperModeApplicable: false,
            isHostedCheckoutEnabled: true,
            isInlineCheckoutEnabled: false,
            isPayPalCreditAvailable: true,
            isVenmoEnabled: true,
            merchantId: 'XdasdQWe1123',
            orderId: undefined,
            shouldRenderFields: true,
            showOnlyOnMobileDevices: false,
        },
        initializationStrategy: {
            type: 'not_applicable',
        },
        logoUrl: 'http://logo_url_path',
        method: 'paypal',
        nonce: undefined,
        returnUrl: undefined,
        supportedCards: [
            'PAYPAL',
            'PAYPALCREDIT',
            'VISA',
            'MC',
            'AMEX',
            'DISCOVER',
            'VENMO',
            'APPLEPAY',
        ],
        type: 'PAYMENT_TYPE_API',
    };
}

export function getPayPalCommerceCreditMethod() {
    const paypalCommerce = getPayPalCommerceMethod();

    return {
        ...paypalCommerce,
        id: 'paypalcommercecredit',
        initializationData: {
            ...paypalCommerce.initializationData,
            payPalCreditProductBrandName: {
                credit: 'Pay Later',
                messaging: 'Pay Later',
            },
        },
        method: 'paypal-credit',
    };
}

export function getPayPalCommerceVenmoMethod() {
    const paypalCommerce = getPayPalCommerceMethod();

    return {
        ...paypalCommerce,
        id: 'paypalcommercevenmo',
        method: 'paypal-venmo',
    };
}

export function getPayPalCommerceAPMsMethod() {
    const paypalCommerce = getPayPalCommerceMethod();

    const apms = ['bancontact', 'eps', 'giropay', 'ideal', 'mybank', 'p24', 'sepa', 'sofort'];

    return {
        ...paypalCommerce,
        config: {
            ...paypalCommerce.config,
            displayName: 'Sepa',
        },
        gateway: 'paypalcommercealternativemethods',
        id: 'sepa',
        initializationData: {
            ...paypalCommerce.initializationData,
            availableAlternativePaymentMethods: apms,
            enabledAlternativePaymentMethods: apms,
        },
        method: 'multi-option',
    };
}
