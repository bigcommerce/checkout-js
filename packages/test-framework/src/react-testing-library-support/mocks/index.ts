export * from './checkout.mock';
export * from './checkout-settings.mock';
export { countries } from './countries';
export { formFields } from './form-fields';
export { payments } from './payments';
