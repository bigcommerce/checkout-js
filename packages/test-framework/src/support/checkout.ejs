<!DOCTYPE html>
<html>
    <head>
        <title>Local Testing Environment</title>
        <base href="/" target="_top"/>
        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
        <link href="https://fonts.googleapis.com/css?family=Karla:400|Montserrat:400,500,700&display=swap" rel="stylesheet" />
    </head>
    <body>
    <header class="checkoutHeader optimizedCheckout-header">
        <div class="checkoutHeader-content">
            <h1 class="is-srOnly">Checkout</h1>
            <h2 class="checkoutHeader-heading">
                <a class="checkoutHeader-link" href="<%= storeUrl %>/">
                    <span class="header-logo-text">Local Testing Environment</span>
                </a>
            </h2>
        </div>
    </header>
    <div class="ng-checkout-container" id="micro-app-ng-checkout">
        <div class="ng-checkout-view" data-ui-view></div>
    </div>
    <script type="text/javascript">
        window.checkoutConfig = {
            checkoutId: '<%= checkoutId %>',
            containerId: 'micro-app-ng-checkout',
        };
    </script>
    <script src="<%= localhostUrl %>/auto-loader.js" type="text/javascript"></script>
    <script type="text/javascript">
        (function() {
            window.checkout.renderCheckout({
                checkoutId: '<%= checkoutId %>',
                containerId: 'micro-app-ng-checkout',
                sentryConfig: null,
            })
        })();
    </script>
    </body>
</html>
