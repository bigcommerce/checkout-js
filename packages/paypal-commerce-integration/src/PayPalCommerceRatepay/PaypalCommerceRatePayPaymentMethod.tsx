import { FormField } from '@bigcommerce/checkout-sdk';
import React, { FunctionComponent, useCallback, useEffect, useMemo, useRef } from 'react';

import {
    CountryData,
    CustomError,
    getCountryData,
    PaymentMethodProps,
    PaymentMethodResolveId,
    SpecificError,
    toResolvableComponent,
} from '@bigcommerce/checkout/payment-integration-api';
import { DynamicFormField, DynamicFormFieldType, FormContext } from '@bigcommerce/checkout/ui';

import getPaypalCommerceRatePayValidationSchema from './validation-schema/getPaypalCommerceRatePayValidationSchema';

const PAYMENT_SOURCE_INFO_CANNOT_BE_VERIFIED = 'PAYMENT_SOURCE_INFO_CANNOT_BE_VERIFIED';
const PAYMENT_SOURCE_DECLINED_BY_PROCESSOR = 'PAYMENT_SOURCE_DECLINED_BY_PROCESSOR';
const ITEM_CATEGORY_NOT_SUPPORTED_BY_PAYMENT_SOURCE =
    'ITEM_CATEGORY_NOT_SUPPORTED_BY_PAYMENT_SOURCE';

interface RatePayFieldValues {
    ratepayBirthDate: {
        getDate(): number;
        getMonth(): number;
        getFullYear(): number;
    };
    ratepayPhoneCountryCode: string;
    ratepayPhoneNumber: string;
}

const formFieldData: FormField[] = [
    {
        name: 'ratepayBirthDate',
        custom: false,
        id: 'ratepayBirthDate',
        label: 'payment.ratepay.birth_date',
        required: true,
        fieldType: DynamicFormFieldType.DATE,
        inputDateFormat: 'dd.MM.yyyy',
    },
    {
        name: 'ratepayPhoneCountryCode',
        custom: false,
        id: 'ratepayPhoneCountryCode',
        label: 'payment.ratepay.phone_country_code',
        required: true,
        fieldType: DynamicFormFieldType.TEXT,
        type: 'string',
    },
    {
        name: 'ratepayPhoneNumber',
        custom: false,
        id: 'ratepayPhoneNumber',
        label: 'payment.ratepay.phone_number',
        required: true,
        fieldType: DynamicFormFieldType.TEXT,
    },
];

const PaypalCommerceRatePayPaymentMethod: FunctionComponent<PaymentMethodProps> = ({
    method,
    checkoutService,
    onUnhandledError,
    paymentForm: { isSubmitted, setFieldValue, setValidationSchema, setSubmitted },
    language,
    checkoutState,
}) => {
    const fieldsValues = useRef<Partial<RatePayFieldValues>>({});
    const isPaymentDataRequired = checkoutState.data.isPaymentDataRequired();

    const getCountryInfo = (): CountryData | undefined => {
        const billing = checkoutState.data.getBillingAddress();

        if (billing && !billing.country) {
            return getCountryData(billing.country)[0];
        }
    };

    const initializePayment = async () => {
        try {
            await checkoutService.initializePayment({
                gatewayId: method.gateway,
                methodId: method.id,
                paypalcommerceratepay: {
                    container: '#checkout-payment-continue',
                    legalTextContainer: 'legal-text-container',
                    loadingContainerId: 'checkout-page-container',
                    getFieldsValues: () => {
                        // TODO: update PPCP Ratepay initialization data with getFieldsValues optional return data in checkout-sdk
                        const defaultFieldValues = {
                            ratepayBirthDate: {
                                getDate: () => 0,
                                getMonth: () => 0,
                                getFullYear: () => 0,
                            },
                            ratepayPhoneCountryCode: '',
                            ratepayPhoneNumber: '',
                        };

                        return {
                            ...defaultFieldValues,
                            ...fieldsValues.current,
                        };
                    },
                    onError: (error: SpecificError) => {
                        const ratepaySpecificError = error.errors?.filter((e) => e.provider_error);

                        if (ratepaySpecificError?.length) {
                            let translationCode;
                            let ratepayError;
                            const ratepaySpecificErrorCode =
                                ratepaySpecificError[0].provider_error?.code;

                            switch (ratepaySpecificErrorCode) {
                                case PAYMENT_SOURCE_DECLINED_BY_PROCESSOR:
                                    translationCode =
                                        'payment.ratepay.errors.paymentSourceDeclinedByProcessor';
                                    break;

                                case PAYMENT_SOURCE_INFO_CANNOT_BE_VERIFIED:
                                    translationCode =
                                        'payment.ratepay.errors.paymentSourceInfoCannotBeVerified';
                                    break;

                                case ITEM_CATEGORY_NOT_SUPPORTED_BY_PAYMENT_SOURCE:
                                    translationCode =
                                        'payment.ratepay.errors.itemCategoryNotSupportedByPaymentSource';
                                    break;

                                default:
                                    translationCode = 'common.error_heading';
                            }

                            if (
                                ratepaySpecificErrorCode !==
                                ITEM_CATEGORY_NOT_SUPPORTED_BY_PAYMENT_SOURCE
                            ) {
                                ratepayError = new CustomError({
                                    data: {
                                        shouldBeTranslatedAsHtml: true,
                                        translationKey: translationCode,
                                    },
                                });
                            } else {
                                ratepayError = new Error(language.translate(translationCode));
                            }

                            return onUnhandledError(ratepayError);
                        }

                        onUnhandledError(error);
                    },
                },
            });
        } catch (error) {
            if (error instanceof Error) {
                onUnhandledError(error);
            }
        }
    };

    const deinitializePayment = async () => {
        try {
            await checkoutService.deinitializePayment({
                gatewayId: method.gateway,
                methodId: method.id,
            });
        } catch (error) {
            if (error instanceof Error) {
                onUnhandledError(error);
            }
        }
    };

    useEffect(() => {
        void initializePayment();

        return () => {
            void deinitializePayment();
        };
    }, []);

    const updateFieldValues = (field: { [key: string]: string }): void => {
        fieldsValues.current = { ...fieldsValues.current, ...field };
    };

    const handleChange = useCallback(
        (fieldId: string) => (value: string) => {
            setFieldValue(fieldId, value);
            updateFieldValues({ [fieldId]: value });
        },
        [setFieldValue],
    );

    const validationSchema = useMemo(
        () =>
            getPaypalCommerceRatePayValidationSchema({
                formFieldData,
                language,
            }),
        [language, formFieldData],
    );

    useEffect(() => {
        setSubmitted(false);
        setValidationSchema(method, validationSchema);
    }, [validationSchema, method, setValidationSchema, setSubmitted]);

    useEffect(() => {
        const countryInfo = getCountryInfo();

        if (countryInfo) {
            setFieldValue('ratepayPhoneCountryCode', countryInfo.dialCode);
        }
    }, []);

    if (!isPaymentDataRequired) {
        return null;
    }

    const formContextProps = {
        isSubmitted: isSubmitted(),
        setSubmitted,
    };

    return (
        <div style={{ marginBottom: '20px' }}>
            <FormContext.Provider value={formContextProps}>
                {formFieldData.map((field) => {
                    return (
                        <DynamicFormField
                            extraClass={`dynamic-form-field--${field.id}`}
                            field={field}
                            key={field.id}
                            label={language.translate(field.label)}
                            onChange={handleChange(field.id)}
                        />
                    );
                })}
            </FormContext.Provider>
        </div>
    );
};

export default toResolvableComponent<PaymentMethodProps, PaymentMethodResolveId>(
    PaypalCommerceRatePayPaymentMethod,
    [{ gateway: 'paypalcommercealternativemethods', id: 'ratepay' }],
);
