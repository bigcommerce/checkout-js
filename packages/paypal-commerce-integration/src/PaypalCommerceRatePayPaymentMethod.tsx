import React, { FunctionComponent, useCallback, useEffect, useMemo, useRef, useState } from 'react';

import {
    PaymentMethodProps,
    PaymentMethodResolveId,
    toResolvableComponent,
} from '@bigcommerce/checkout/payment-integration-api';
import { DynamicFormField, DynamicFormFieldType, FormContext } from '@bigcommerce/checkout/ui';
import { FormField } from '@bigcommerce/checkout-sdk';
import getPaypalCommerceRatePayValidationSchema from './validation-schemas/getPaypalCommerceRatePayValidationSchema';
import { LoadingSpinner } from '@bigcommerce/checkout/ui';

interface RatePayFieldValues {
    ratepayBirthDate: {
        getDate(): number;
        getMonth(): number;
        getFullYear(): number;
    };
    ratepayPhoneCountryCode: string;
    ratepayPhoneNumber: string;
}

const formFieldData: FormField[] = [
    {
        name: 'ratepayBirthDate',
        custom: false,
        id: 'ratepayBirthDate',
        label: 'payment.ratepay.birth_date',
        required: true,
        fieldType: DynamicFormFieldType.DATE,
    },
    {
        name: 'ratepayPhoneCountryCode',
        custom: false,
        id: 'ratepayPhoneCountryCode',
        label: 'payment.ratepay.phone_country_code',
        required: true,
        fieldType: DynamicFormFieldType.TEXT,
        type: 'string',
        maxLength: 2,
    },
    {
        name: 'ratepayPhoneNumber',
        custom: false,
        id: 'ratepayPhoneNumber',
        label: 'payment.ratepay.phone_number',
        required: true,
        fieldType: DynamicFormFieldType.TEXT,
        maxLength: 11,
        min: 8,
    }
];

const PaypalCommerceRatePayPaymentMethod: FunctionComponent<any> = ({
    method,
    checkoutService,
    paymentForm,
    onUnhandledError,
    paymentForm: {
        isSubmitted,
        setFieldValue,
        setValidationSchema,
        setSubmitted,
    },
    language,
    checkoutState,
}) => {
    const fieldsValues = useRef<Partial<RatePayFieldValues>>({});
    const isPaymentDataRequired = checkoutState.data.isPaymentDataRequired();
    const [isPaymentSubmitting, setIsPaymentSubmitting] = useState(false);

    if (!isPaymentDataRequired) {
        return null;
    }

    const initializePayment = async () => {
        try {
            await checkoutService.initializePayment({
                gatewayId: method.gateway,
                methodId: method.id,
                paypalcommerceratepay: {
                    container: '#checkout-payment-continue',
                    legalTextContainer: 'legal-text-container',
                    getFieldsValues: () => fieldsValues.current,
                    onPaymentSubmission: (isSubmitting: boolean) => setIsPaymentSubmitting(isSubmitting),
                    onError: (error: Error) => {
                        paymentForm.disableSubmit(method, true);
                        onUnhandledError(error);
                    },
                },
            });
        } catch (error) {
            if (error instanceof Error) {
                onUnhandledError(error);
            }
        }
    };

    const deinitializePayment = async () => {
        try {
            await checkoutService.deinitializePayment({
                gatewayId: method.gateway,
                methodId: method.id,
            });
        } catch (error) {
            if (error instanceof Error) {
                onUnhandledError(error);
            }
        }
    };

    useEffect(() => {
        void initializePayment();

        return () => {
            void deinitializePayment();
        };
    }, []);

    const handleChange = useCallback(
        (fieldId: string) => (value: string) => {
            setFieldValue(fieldId, value);
            updateFieldValues({ [fieldId]: value });
        },
        [setFieldValue],
    );

    const updateFieldValues = (field: { [key: string]: string }) => {
        fieldsValues.current = {...fieldsValues.current, ...field};
    }

    const validationSchema = useMemo(() =>
            getPaypalCommerceRatePayValidationSchema({
                formFieldData,
                language,
            })
        , [language, formFieldData]);

    useEffect(() => {
        setSubmitted(false);
        setValidationSchema(method, validationSchema);
    }, [validationSchema, method, setValidationSchema, setSubmitted]);

    return (
        <div style={{marginBottom:'20px'}}>
            { isPaymentSubmitting &&
                <div className='embedded-checkout-loading-spinner-overlay'>
                    <LoadingSpinner isLoading={true}/>
                </div>
            }
            <FormContext.Provider value={{ isSubmitted, setSubmitted }}>
                { formFieldData.map((field) => {
                        return  <DynamicFormField
                            extraClass={`dynamic-form-field--${field.id}`}
                            field={field}
                            key={field.id}
                            label={language.translate(field.label)}
                            onChange={handleChange(field.id)}
                        />
                    }
                )
                }
            </FormContext.Provider>
        </div>
    )
};

export default toResolvableComponent<PaymentMethodProps, PaymentMethodResolveId>(
    PaypalCommerceRatePayPaymentMethod,
    [{ gateway: 'paypalcommercealternativemethods', id:'ratepay' }],
);
