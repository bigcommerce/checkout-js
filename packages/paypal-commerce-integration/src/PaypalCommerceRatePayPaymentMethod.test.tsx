import { createCheckoutService, LanguageService } from '@bigcommerce/checkout-sdk';
import { fireEvent, render } from '@testing-library/react';
import { Formik } from 'formik';
import React, { FunctionComponent } from 'react';

import {
    createLocaleContext,
    LocaleContext,
    LocaleContextType,
} from '@bigcommerce/checkout/locale';
import {
    PaymentFormService,
    PaymentMethodProps,
} from '@bigcommerce/checkout/payment-integration-api';
import { getCheckout as getCheckoutMock, getStoreConfig } from '@bigcommerce/checkout/test-mocks';
import { FormContext } from '@bigcommerce/checkout/ui';

import { getPaypalCommerceRatePayMethodMock } from './mocks/paypalCommerceRatePayMocks';
import PaypalCommerceRatePayPaymentMethod from './PaypalCommerceRatePayPaymentMethod';

describe('PaypalCommerceRatePayPaymentMethod', () => {
    let PaypalCommerceRatePayPaymentMethodTest: FunctionComponent<PaymentMethodProps>;
    let localeContext: LocaleContextType;
    const checkoutService = createCheckoutService();
    const checkoutState = checkoutService.getState();
    const props = {
        method: getPaypalCommerceRatePayMethodMock(),
        checkoutService,
        checkoutState,
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        paymentForm: {
            setSubmitted: jest.fn(),
            setValidationSchema: jest.fn(),
            setFieldValue: jest.fn(),
        } as unknown as PaymentFormService,
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        language: { translate: jest.fn() } as unknown as LanguageService,
        onUnhandledError: jest.fn(),
    };

    beforeEach(() => {
        jest.spyOn(checkoutState.data, 'getCheckout').mockReturnValue(getCheckoutMock());
        jest.spyOn(checkoutState.data, 'isPaymentDataRequired').mockReturnValue(true);
        localeContext = createLocaleContext(getStoreConfig());

        const submit = jest.fn();

        PaypalCommerceRatePayPaymentMethodTest = (props: PaymentMethodProps) => (
            <LocaleContext.Provider value={localeContext}>
                <FormContext.Provider value={{ isSubmitted: true, setSubmitted: jest.fn() }}>
                    <Formik initialValues={{}} onSubmit={submit}>
                        {({ handleSubmit }) => (
                            <form aria-label="form" onSubmit={handleSubmit}>
                                <PaypalCommerceRatePayPaymentMethod {...props} />
                            </form>
                        )}
                    </Formik>
                </FormContext.Provider>
            </LocaleContext.Provider>
        );
    });

    it('successfully initializes payment with required props', () => {
        const initializePayment = jest
            .spyOn(checkoutService, 'initializePayment')
            .mockResolvedValue(checkoutState);

        render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        expect(initializePayment).toHaveBeenCalledWith({
            gatewayId: props.method.gateway,
            methodId: props.method.id,
            paypalcommerceratepay: {
                onPaymentSubmission: expect.any(Function),
                container: '#checkout-payment-continue',
                legalTextContainer: 'legal-text-container',
                getFieldsValues: expect.any(Function),
                onError: expect.any(Function),
            },
        });
    });

    it('renders component with required fields', async () => {
        const view = render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        expect(view).toMatchSnapshot();
    });

    it('submits form', async () => {
        render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        const submitForm = jest.fn();
        const form = document.querySelectorAll('form')[0];

        form.onsubmit = submitForm;

        fireEvent.submit(form);

        await new Promise((resolve) => process.nextTick(resolve));

        expect(submitForm).toHaveBeenCalled();
    });

    it('deinitializes PaypalCommerceRatePayPaymentMethod', () => {
        const deinitializePayment = jest
            .spyOn(checkoutService, 'deinitializePayment')
            .mockResolvedValue(checkoutState);
        const component = render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        component.unmount();

        expect(deinitializePayment).toHaveBeenCalledWith({
            gatewayId: props.method.gateway,
            methodId: props.method.id,
        });
    });

    it('catches error during PaypalCommerceRatePayPaymentMethod initialization', async () => {
        jest.spyOn(checkoutService, 'initializePayment').mockRejectedValue(new Error('test error'));
        render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        await new Promise((resolve) => process.nextTick(resolve));

        expect(props.onUnhandledError).toHaveBeenCalled();
    });

    it('catches error during PaypalCommerceRatePayPaymentMethod deinitialization', async () => {
        jest.spyOn(checkoutService, 'deinitializePayment').mockRejectedValue(
            new Error('test error'),
        );

        const component = render(<PaypalCommerceRatePayPaymentMethodTest {...props} />);

        await new Promise((resolve) => process.nextTick(resolve));

        component.unmount();

        await new Promise((resolve) => process.nextTick(resolve));

        expect(props.onUnhandledError).toHaveBeenCalled();
    });
});
