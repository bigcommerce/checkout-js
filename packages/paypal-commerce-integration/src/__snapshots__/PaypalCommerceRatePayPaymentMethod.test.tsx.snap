// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaypalCommerceRatePayPaymentMethod renders component with required fields 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        aria-label="form"
      >
        <div
          style="margin-bottom: 20px;"
        >
          <div
            class="dynamic-form-field dynamic-form-field--ratepayBirthDate"
          >
            <div
              class="form-field"
            >
              <label
                class="form-label optimizedCheckout-form-label"
                for="ratepayBirthDate"
                id="ratepayBirthDate-label"
              >
                payment.ratepay.birth_date
              </label>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container "
                >
                  <span
                    aria-live="polite"
                    class="react-datepicker__aria-live"
                    role="alert"
                  />
                  <input
                    autocomplete="off"
                    class="form-input optimizedCheckout-form-input"
                    name="ratepayBirthDate"
                    placeholder="DD/MM/YYYY"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="dynamic-form-field dynamic-form-field--ratepayPhoneCountryCode"
          >
            <div
              class="form-field"
            >
              <label
                class="form-label optimizedCheckout-form-label"
                for="ratepayPhoneCountryCode"
                id="ratepayPhoneCountryCode-label"
              >
                payment.ratepay.phone_country_code
              </label>
              <input
                aria-labelledby="ratepayPhoneCountryCode-label ratepayPhoneCountryCode-field-error-message"
                class="form-input optimizedCheckout-form-input"
                data-test="ratepayPhoneCountryCode-text"
                id="ratepayPhoneCountryCode"
                maxlength="2"
                name="ratepayPhoneCountryCode"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="dynamic-form-field dynamic-form-field--ratepayPhoneNumber"
          >
            <div
              class="form-field"
            >
              <label
                class="form-label optimizedCheckout-form-label"
                for="ratepayPhoneNumber"
                id="ratepayPhoneNumber-label"
              >
                payment.ratepay.phone_number
              </label>
              <input
                aria-labelledby="ratepayPhoneNumber-label ratepayPhoneNumber-field-error-message"
                class="form-input optimizedCheckout-form-input"
                data-test="ratepayPhoneNumber-text"
                id="ratepayPhoneNumber"
                maxlength="11"
                min="8"
                name="ratepayPhoneNumber"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      aria-label="form"
    >
      <div
        style="margin-bottom: 20px;"
      >
        <div
          class="dynamic-form-field dynamic-form-field--ratepayBirthDate"
        >
          <div
            class="form-field"
          >
            <label
              class="form-label optimizedCheckout-form-label"
              for="ratepayBirthDate"
              id="ratepayBirthDate-label"
            >
              payment.ratepay.birth_date
            </label>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container "
              >
                <span
                  aria-live="polite"
                  class="react-datepicker__aria-live"
                  role="alert"
                />
                <input
                  autocomplete="off"
                  class="form-input optimizedCheckout-form-input"
                  name="ratepayBirthDate"
                  placeholder="DD/MM/YYYY"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="dynamic-form-field dynamic-form-field--ratepayPhoneCountryCode"
        >
          <div
            class="form-field"
          >
            <label
              class="form-label optimizedCheckout-form-label"
              for="ratepayPhoneCountryCode"
              id="ratepayPhoneCountryCode-label"
            >
              payment.ratepay.phone_country_code
            </label>
            <input
              aria-labelledby="ratepayPhoneCountryCode-label ratepayPhoneCountryCode-field-error-message"
              class="form-input optimizedCheckout-form-input"
              data-test="ratepayPhoneCountryCode-text"
              id="ratepayPhoneCountryCode"
              maxlength="2"
              name="ratepayPhoneCountryCode"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="dynamic-form-field dynamic-form-field--ratepayPhoneNumber"
        >
          <div
            class="form-field"
          >
            <label
              class="form-label optimizedCheckout-form-label"
              for="ratepayPhoneNumber"
              id="ratepayPhoneNumber-label"
            >
              payment.ratepay.phone_number
            </label>
            <input
              aria-labelledby="ratepayPhoneNumber-label ratepayPhoneNumber-field-error-message"
              class="form-input optimizedCheckout-form-input"
              data-test="ratepayPhoneNumber-text"
              id="ratepayPhoneNumber"
              maxlength="11"
              min="8"
              name="ratepayPhoneNumber"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
